draft-aegis-mimi-arch-latest.md

```diff
@@ -189,37 +188,99 @@ struct {
 
 ~~~
-struct {
-   uint8 start;
-   uint8 end<0..255>;
-} X509IdentityRange;
-
-fn client_identifier(HashFunction hasher, CertificateChain cert_chain,
-X509IdentityRange range)
+fn identifier(CertificateChain cert_chain, uint8 start, uint8 end)
 {
-    for certificate in cert_chain[range.start..=range.end] {
+    string id = new String;
+    bool first = true;
+    
+    for certificate in cert_chain[start...end] {
         let common_name = certificate.subject.common_name;
 
         if (!common_name) {
             throw "Common name is required";
         }
+        
+        if (common_name.contains("@","/","#","$",":") {
+            throw "Common names may not contain @/#$: characters.";
+        }
 
-        hasher.update(common_name);
+        if (!first) {
+            id.append(":");
+        } else {
+            first = false;
+        }
+        id.append(common_name);
     }
 
-    return hasher.finalize();
+    return id;
+}
+~~~
+
+For example client identifiers are given by fixing a range and calling 'identifier':
+
+~~~
+struct {
+    uint8 start;
+    uint8 end<0...255>;
+} X509ClientIdentityRange
+
+fn client_identifier(CertificateChain cert_chain, X509ClientIdentityRange range) {
+    return identifier(cert_chain, range.start, range.end)
+}
+~~~
+
+### X.509 Account Identifiers
+
+Many messaging systems use of multi-client (e.g. multi-device) accounts. Account identifiers are calculated just like client identifiers but for the Account Identity range.
+
+~~~
+struct {
+    uint8 start;
+    uint8 end<0...255>;
+} X509AccountIdentityRange
+
+fn acount_identifier(CertificateChain cert_chain, X509AccountIdentityRange range) {
+    return identifier(cert_chain, range.start, range.end)
 }
 ~~~
 
+The Account Identity range strictly succeeds the Client Identity range in the certificate chain counting from the leaf certificate up.
+
+~~~
+X509AccountIdentityRange.start > X509ClientIdentityRange.end
+~~~~
+
+### X.509 Domain Identifiers
+
+Federated messaging systems often associate accounts with a domain (e.g. that of a home server hosting the account). Domain identifiers are calculated just like client identifiers but for the Domain Identity range.
+
+~~~
+struct {
+    uint8 start;
+    uint8 end<0...255>;
+} X509DomainIdentityRange
+
+fn domain_identifier(CertificateChain cert_chain, X509DomainIdentityRange range) {
+    return identifier(cert_chain, range.start, range.end)
+}
+~~~
+
+The Domain Identity range strictly succeeds the Client Identity range in the certificate chain counting from the leaf certificate up.
+
+~~~
+X509DomainIdentityRange.start > X509ClientIdentityRange.end
+~~~~
+
+When Account identifiers are used then the Domain Identity range must also strictly succeed the Account Identity range.
+
+~~~
+X509DomainIdentityRange.start > X509AccountIdentityRange.end
+~~~~
```

Comment:
I'm not sure I see the value in having a range for either of these. In practice you must have a single certificate that describes your account / domain name just for practicality. I want to be able to pull messaging.example.org out as a string so I can use it to direct traffic over the internet, and once you have hashing that data is no longer useful. 

Even with HTTPS, your cert must have your entire domain and is not constructed via subdomains in separate certs. 

Ultimately IMO, the interface should just have a single index for account, and single index for domain. Those can be used for constructing UI like tomleavy:mimi.amazon.com or routing traffic to the appropriate domain as needed

Reply:
yup, fair enough. will implement those changes.